### DISCUSSION TITLE
How to test components that make use of the `useEnvironment` hook internally?

### DISCUSSION THREAD

**Discussion by mrmartineau** [chakra-ui/chakra-ui#6664] [Q&A]
FYI I am using Vitest as my test runner, but I think the same would apply to Jest as well.

I have a component that makes use of the `useBreakpointValue` hook. I want to test that when a button is clicked, a `Drawer` opens and reveals some specific content (there's a bit more to it than that, but that is essentially it). I do not need to test Chakra's `Drawer` itself (because that is already well tested) but my logic to open based on button clicks etc.

Internally [`useBreakpointValue`](https://github.com/chakra-ui/chakra-ui/blob/main/packages/components/media-query/src/use-breakpoint.ts) makes use of the [`useMediaQuery`](https://github.com/chakra-ui/chakra-ui/blob/main/packages/components/media-query/src/use-media-query.ts) hook which uses the [`useEnvironment`](https://github.com/chakra-ui/chakra-ui/blob/3a87f68474221ed402fe5d169f7067763c8436d5/packages/components/env/src/env.tsx#L29) hook. 

When I ran my tests, I got this error:

```
Error: Uncaught [TypeError: env.window.matchMedia is not a function]
```

So I realised I needed to mock `window.matchMedia` in Vitest like so:

```js
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: vi.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: vi.fn(),
    removeListener: vi.fn(),
    addEventListener: vi.fn(),
    removeEventListener: vi.fn(),
    dispatchEvent: vi.fn(),
  })),
})
```
With Jest this is what would be needed:
```js
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation(query => {
    return ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    });
  }),
});
```

Unfortunately this did not work because instead of `window.matchMedia` I need to mock `env.window.matchMedia`.. 

The `env` here comes from the `useEnvironment` hook and that is what I don't know how to mock...

I tried this instead:

```js
import { useEnvironment } from '@chakra-ui/react';
import { vi } from 'vitest';

const env = useEnvironment();

Object.defineProperty(env, 'window', {
  writable: true,
  value: {
    matchMedia: vi.fn().mockImplementation((query) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: vi.fn(),
      removeListener: vi.fn(),
      addEventListener: vi.fn(),
      removeEventListener: vi.fn(),
      dispatchEvent: vi.fn(),
    })),
  },
});
```

But it did not work.. because that hook needs to be a "React" context, it revealed these errors:

```
Warning: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
See https://reactjs.org/link/invalid-hook-call for tips about how to debug and fix this problem.

TypeError: Cannot read properties of null (reading 'useContext')
 ❯ useContext node_modules/react/cjs/react.development.js:1618:21
 ❯ Proxy.useEnvironment node_modules/@chakra-ui/react-env/dist/index.cjs.js:141:39
 ❯ vitest.setup.ts:7:12
      7| const env = useEnvironment();
```

Having seen how the [Storybook stories for the env-related](https://github.com/chakra-ui/chakra-ui/blob/3a87f68474221ed402fe5d169f7067763c8436d5/packages/components/env/stories/env.stories.tsx) code are created, I tried my own variation on that, which looks like the code below and wraps the component I am trying to test: 

```tsx
const Wrapper = ({ children }: React.PropsWithChildren<{}>) => {
  const { window, document } = useEnvironment();
  return (
    <EnvironmentProvider environment={{ window, document }}>
      {children}
    </EnvironmentProvider>
  );
};
```

Which also showed the `env.window.matchMedia` error.. 😞. `EnvironmentProvider` looks like it should allow me to do what I need, but it does not..

I can feel that I'm close but I can't quite figure it out. Any ideas?

**Comment by rgpass** [comment]
@mrmartineau Were you able to figure this out? I'm running into the same thing and can't figure it out.

**Comment by tadeo** [comment]
I solve the issue in jest mocking window.matchMedia like this:
```javascript
// utils.js
export function mockMatchMedia() {
  Object.defineProperty(window, "matchMedia", {
    value: jest.fn(() => ({ matches: true, addListener: jest.fn(), removeListener: jest.fn() }))
  });
}
```
```javascript
// test.js
import { mockMatchMedia } from './utils';

describe('chakra ui component testing', () => {
  beforeAll(mockMatchMedia);

  it('renders', () => {
    // test a component wrapped with ChakraProvider containing Chakra UI components
    //  render(
    //    <ChakraProvider theme={theme}>
    //      <ComponentWithChakraUIComponents />
    //    </ChakraProvider>
    //  );
  });
});
```
I hope this helps!

**Comment by lakardion** [comment]
I'm having this problem also. Need to mock `env.window` and `useEnvironment` is not helping with that.

Keep getting the `env.window.matchMedia` not defined.

Simply mocking window with `Object.defineProperty(window, "matchMedia", {...` is not getting it there, same as what OP showcased

**Comment by andersonfpcorrea** [comment]
I was able to solve this using `vi.importActual` as follows:
```javascript
// Mock Chakra's useBreakpointValue hook:
vi.mock("@chakra-ui/react", async () => {
  const mod = await vi.importActual("@chakra-ui/react");
  return {
    ...(mod as Record<string, unknown>),
    useBreakpointValue: vi.fn().mockImplementation(() => "sm"),
  };
});
```

**Reply by ugogo** [reply]
Thanks @andersonfpcorrea, this worked for me too 🙏

**Comment by andersonfpcorrea** [comment]
Just to add another approach, instead of mocking ```useBreakpointValue``` hook as I mentioned above, we can directly mock ```window.matchMedia``` by using ```vi.stubGlobal```, as mentioned in [Vitest documentation](https://vitest.dev/guide/mocking.html#globals)

```javascript
const matchMediaMock = vi.fn((query: unknown) => ({
  matches: false,
  media: query,
  onchange: null,
  addListener: vi.fn(),
  removeListener: vi.fn(),
  addEventListener: vi.fn(),
  removeEventListener: vi.fn(),
  dispatchEvent: vi.fn(),
}));
vi.stubGlobal("matchMedia", matchMediaMock);
```

I recently used this way of mocking `matchMedia` for testing Ant D components

**Reply by szympajka** [reply]
Best answer! `Object.defineProperty` worked as mock for me, but when I tried to read from the value of `mediaQuery`, it was `undefined` — using `vi.stubGlobal` fixed my issue :)

**Comment by gbxbro** [comment]
After fiddling with this crap for about a week I solved the riddle

```javascript
const ChakraWrapper = ({
  children,
}: {
  children: ReactElement
}) => {
  const env = useEnvironment();

  Object.defineProperty(env, 'window', {
    writable: true,
    value: {
      matchMedia: vi.fn().mockImplementation((query) => {
        return {
          matches: false,
          media: query,
          onchange: null,
          addListener: vi.fn(),
          removeListener: vi.fn(),
          addEventListener: vi.fn(),
          removeEventListener: vi.fn(),
          dispatchEvent: vi.fn(),
        };
      }),
    },
  });
  return (
    <ChakraProvider theme={theme} environment={env}>
      {children}
    </ChakraProvider>
  );
};

it('renders YourComponent', async () => {
    render(<YourComponent />, {
      wrapper: ChakraWrapper,
    });
  });
```

**Reply by gbxbro** [reply]
if you use, in addition to Chakra, also @emotion/styled, then 

```javascript
vi.mock("@chakra-ui/react")
```

will cause you a number of other problems with tests

**Reply by gbxbro** [reply]
I used useEnvironment for typization, cause interface not exported from chakra. You can use default JS object
